lms/djangoapps/certificates/tasks.py

```diff
@@ -18,11 +19,30 @@ class _BaseCertificateTask(PersistOnFailureTask, LoggedTask):  # pylint: disable
     abstract = True
 
 
-@task(base=_BaseCertificateTask)
-def generate_certificate(**kwargs):
+@task(base=_BaseCertificateTask, bind=True, default_retry_delay=30, max_retries=2)
+def generate_certificate(self, **kwargs):
     """
     Generates a certificate for a single user.
+
+    kwargs:
+        - student: The student for whom to generate a certificate.
+        - course_key: The course key for the course that the student is
+            receiving a certificate in.
+        - expected_verification_status: The expected verification status
+            for the user.  When the status has changed, we double check
+            that the actual verification status is as expected before
+            generating a certificate, in the off chance that the database
+            has not yet updated with the user's new verification status.
+            status.
     """
     student = User.objects.get(id=kwargs.pop('student'))
     course_key = CourseKey.from_string(kwargs.pop('course_key'))
-    generate_user_certificates(student=student, course_key=course_key, **kwargs)
+    expected_verification_status = kwargs.pop('expected_verification_status', None)
+    if not expected_verification_status:
+        generate_user_certificates(student=student, course_key=course_key, **kwargs)
+    else:
+        actual_verification_status, _ = SoftwareSecurePhotoVerification.user_status(student)
+        if expected_verification_status == actual_verification_status:
+            generate_user_certificates(student=student, course_key=course_key, **kwargs)
+        else:
+            self.retry(kwargs=kwargs)
```

Comment:
You can both DRY and simplify this logic by replacing the above with:

```
expected_verification_status = kwargs.pop('expected_verification_status', None)
if expected_verification_status:
      actual_verification_status, _ = SoftwareSecurePhotoVerification.user_status(student)
      if expected_verification_status != actual_verification_status:
          raise self.retry(kwargs=kwargs)
generate_user_certificates(student=student, course_key=course_key, **kwargs)
```

Reply:
We had this before, but got tripped up because we weren't sure if generate_user_certificates would be called after self.retry or not (it doesn't, because self.retry raises an exception, but I don't think it's obvious from the code that that's the case). @iloveagent57 and I were trying to figure out which way was cleaner and (begrudgingly) settled on the other way, but with the "raise" written explicitly I think this is better. :)